crates/hir-ty/src/infer/unify.rs

```diff
@@ -898,6 +899,33 @@ impl<'a> InferenceTable<'a> {
             _ => c,
         }
     }
+
+    /// Check if given type is `Sized` or not
+    pub(crate) fn is_sized(&mut self, ty: &Ty) -> bool {
+        if let Some((AdtId::StructId(id), subst)) = ty.as_adt() {
+            let struct_data = self.db.struct_data(id);
+            if let Some((last_field, _)) = struct_data.variant_data.fields().iter().last() {
+                let last_field_ty =
+                    self.db.field_types(id.into())[last_field].clone().substitute(Interner, subst);
+                // Structs can have DST as its last field and such cases are not handled
+                // as unsized by the chalk, so we do this manually
```

Comment:
(not sure how relevant this is but) this reminds me, we do not populate chalk's `AdtvariantDatum`'s fields  right now as that slows down things by quite a bit. Just letting you know so you are aware (the relevant part for that is stubbed out at the moment even though we have it implemented) https://github.com/rust-lang/rust-analyzer/blob/134e0a4e874e7a739b4e96c987c6b520f8da37ff/crates/hir-ty/src/chalk_db.rs#L768-L783

Reply:
Oh, I was wondering why chalk is saying `struct Foo { inner: [u8] }` is `Sized` and that was the answer😄

Reply:
iirc it added like 10 seconds to inference on self which is a bit much + at the time we couldnt inspect third party deps of the standard library (which we soon can) so we could reconsider if this turns out to cause troubles

Reply:
In that case we might omit it and consider it as sized; that will make some false negative cast errors but we'll not have irritating false positives anyway (though we should change function name into somewhat like `maybe_sized` to avoid possible footguns in future that might be caused by trusting that function will always return the correct value 😅 )